File: database/model/tunnel.go
package model

// TunnelServer and TunnelClient are the merged home of the two tunnel flavours
// the panel manages. AmneziaWG and native WireGuard were kept in four tables
// (awg_servers/awg_clients/wg_servers/wg_clients) whose columns turned out to
// be identical except for AmneziaWG's twelve obfuscation fields, which the
// WireGuard rows simply leave at zero.
//
// Kind ("awg" / "wg") lives on the server; a client inherits its flavour
// through ServerId. The unique index on Kind preserves today's rule of one
// server per flavour — lift it deliberately if multiple tunnels per flavour
// ever become a feature.

// TunnelServer is one tunnel interface with its settings.
type TunnelServer struct {
	// Kind selects the flavour: TunnelKindAwg or TunnelKindWg.
	Kind string `json:"kind" gorm:"uniqueIndex;not null"`

	Id int `json:"id" gorm:"primaryKey;autoIncrement"`

	Enable        bool   `json:"enable" gorm:"default:false"`
	InterfaceName string `json:"interfaceName"`
	ListenPort    int    `json:"listenPort" gorm:"default:51820"`
	MTU           int    `json:"mtu" gorm:"default:1420"`

	// Server keys
	PrivateKey string `json:"privateKey"`
	PublicKey  string `json:"publicKey"`

	// IPv4 tunnel network
	IPv4Address string `json:"ipv4Address" gorm:"default:'10.66.66.1/24'"`
	IPv4Pool    string `json:"ipv4Pool" gorm:"default:'10.66.66.0/24'"`

	// IPv6 — native public addresses
	IPv6Enabled bool   `json:"ipv6Enabled" gorm:"default:false"`
	IPv6Address string `json:"ipv6Address"` // server address on awg0, e.g. "2a01:xxx::1/112"
	IPv6Pool    string `json:"ipv6Pool"`    // pool for clients, e.g. "2a01:xxx::/112"
	IPv6Gateway string `json:"ipv6Gateway"` // upstream gateway for NDP

	// AmneziaWG obfuscation parameters (1.x core set: Jc/Jmin/Jmax/S1/S2/H1-H4).
	Jc   int `json:"jc"`
	Jmin int `json:"jmin"`
	Jmax int `json:"jmax"`
	S1   int `json:"s1"`
	S2   int `json:"s2"`
	// AmneziaWG 2.0 additions. S3/S4 add padding to cookie/transport packets;
	// I1 is a CPS signature packet sent before handshakes. H1-H4 are stored as
	// strings so each holds either a single 1.x value ("1") or a 2.0 range
	// ("100000-800000"). Empty S3/S4/I1 ⇒ the server stays on classic 1.x output.
	S3 int    `json:"s3"`
	S4 int    `json:"s4"`
	H1 string `json:"h1"`
	H2 string `json:"h2"`
	H3 string `json:"h3"`
	H4 string `json:"h4"`
	I1 string `json:"i1"`

	// DNS pushed to clients, split by family. Composed into one DNS line in the
	// client config; the IPv6 entry is used only when IPv6 is enabled.
	DnsIpv4 string `json:"dnsIpv4" gorm:"default:'1.1.1.1'"`
	DnsIpv6 string `json:"dnsIpv6" gorm:"default:'2606:4700:4700::1111'"`

	// External interface for NAT (IPv4)
	ExternalInterface string `json:"externalInterface" gorm:"default:''"`

	// External interface for NDP proxy / IPv6 forwarding (may differ from IPv4)
	IPv6ExternalInterface string `json:"ipv6ExternalInterface" gorm:"default:''"`

	// PostUp / PostDown scripts (auto-generated but overridable)
	PostUp   string `json:"postUp"`
	PostDown string `json:"postDown"`

	// Endpoint that clients connect to (server public IP/domain)
	Endpoint string `json:"endpoint"`

	// Periodic traffic reset: never, daily, weekly, monthly
	TrafficReset string `json:"trafficReset" gorm:"default:'never'"`

	// Route tunnel traffic into Xray via a dokodemo-door TPROXY inbound.
	// When enabled, PostUp installs mangle/TPROXY rules and policy routing
	// that redirect awg0 ingress to a loopback Xray inbound tagged
	// XrayInboundTag, and skips the default MASQUERADE. Routing decisions
	// (which outbound to chain to) are then made by Xray's routing rules.
	RouteViaXray   bool   `json:"routeViaXray" gorm:"default:false"`
	XrayInboundTag string `json:"xrayInboundTag"`
	XrayTproxyPort int    `json:"xrayTproxyPort"`

	CreatedAt int64 `json:"createdAt" gorm:"autoCreateTime:milli"`
	UpdatedAt int64 `json:"updatedAt" gorm:"autoUpdateTime:milli"`
}

// TunnelClient is one peer of a TunnelServer.
//
// Note for whoever switches the services over: Enable deliberately has no gorm
// default. The legacy models used default:true, which makes gorm drop an
// explicit false on Create — a client added as disabled came out enabled. The
// flip side is that new clients must now have Enable set explicitly by the
// service instead of inheriting it from the schema.
type TunnelClient struct {
	Id       int `json:"id" gorm:"primaryKey;autoIncrement"`
	ServerId int `json:"serverId" gorm:"index;uniqueIndex:idx_tunnel_client_server_email,priority:1"`

	UUID    string `json:"uuid" gorm:"uniqueIndex"`
	Name    string `json:"name"`
	Email   string `json:"email" gorm:"uniqueIndex:idx_tunnel_client_server_email,priority:2"`
	Enable  bool   `json:"enable" gorm:"index:idx_tunnel_enable_last_online,priority:1"` // no gorm default: default:true makes Create() drop an explicit false
	Comment string `json:"comment"`

	// Client keys
	PrivateKey   string `json:"privateKey"`
	PublicKey    string `json:"publicKey"`
	PresharedKey string `json:"presharedKey"`

	// Allocated addresses
	IPv4Address string `json:"ipv4Address"` // e.g. "10.66.66.2/32"
	IPv6Address string `json:"ipv6Address"` // e.g. "2a01:xxx::2/128"

	// AllowedIPs on server side (what to route to this client)
	AllowedIPs string `json:"allowedIPs"`

	// AllowedIPs on client side (what to route through tunnel)
	ClientAllowedIPs string `json:"clientAllowedIPs" gorm:"default:'0.0.0.0/0,::/0'"`

	// Comma/semicolon-separated list of ports (or ranges like 8000-8100) to DNAT
	// from the server's external interface to this client. Empty = no forwarding.
	ForwardedPorts string `json:"forwardedPorts" gorm:"default:''"`

	PersistentKeepalive int `json:"persistentKeepalive" gorm:"default:25"`

	// Traffic stats. Upload/Download accumulate as lifetime totals (resettable via
	// reset-traffic) and AllTime is the absolute lifetime — same model as VLESS.
	Upload   int64 `json:"upload" gorm:"default:0"`
	Download int64 `json:"download" gorm:"default:0"`
	TotalGB  int64 `json:"totalGB" gorm:"default:0"` // traffic limit in bytes (0 = unlimited); UI stores bytes, compare against Upload+Download
	AllTime  int64 `json:"allTime" gorm:"default:0"`

	// Raw kernel per-peer transfer counters seen at the last poll. They are the
	// baseline for computing lifetime deltas, so Upload/Download survive an
	// interface bounce (the kernel resets per-peer counters to zero on bounce).
	// Internal bookkeeping — not exposed via JSON.
	LastPeerUp   int64 `json:"-" gorm:"default:0"`
	LastPeerDown int64 `json:"-" gorm:"default:0"`

	ExpiryTime int64 `json:"expiryTime" gorm:"default:0"` // 0 = never
	Reset      int   `json:"reset" gorm:"default:0"`      // auto-renew interval in days, 0 = disabled

	LimitIp    int    `json:"limitIp" gorm:"default:0"`                                                   // max simultaneous IPs, 0 = unlimited
	TgId       int64  `json:"tgId" gorm:"default:0"`                                                      // Telegram chat ID for notifications
	LastOnline int64  `json:"lastOnline" gorm:"default:0;index:idx_tunnel_enable_last_online,priority:2"` // last handshake timestamp (ms)
	LastIP     string `json:"lastIp" gorm:"default:''"`                                                   // last known endpoint IP

	CreatedAt int64 `json:"createdAt" gorm:"autoCreateTime:milli"`
	UpdatedAt int64 `json:"updatedAt" gorm:"autoUpdateTime:milli"`
}

const (
	TunnelKindAwg = "awg"
	TunnelKindWg  = "wg"
)

File: web/assets/js/model/tunnel.js
// One model for both tunnel flavours the panel manages. AmneziaWG and native
// WireGuard describe the same server; AmneziaWG adds obfuscation and the two
// use different defaults so they can run side by side on one host.
const TUNNEL_KIND_DEFAULTS = {
    awg: {
        interfaceName: 'awg0',
        listenPort: 51820,
        ipv4Address: '10.66.66.1/24',
        ipv4Pool: '10.66.66.0/24',
        xrayInboundTag: 'awg-tproxy-in',
        xrayTproxyPort: 12345,
        obfuscation: true,
    },
    wg: {
        interfaceName: 'wg0',
        listenPort: 51821,
        ipv4Address: '10.77.77.1/24',
        ipv4Pool: '10.77.77.0/24',
        xrayInboundTag: 'wg-tproxy-in',
        xrayTproxyPort: 12346,
        obfuscation: false,
    },
};

class TunnelServer {
    constructor(kind, data = {}) {
        const d = TUNNEL_KIND_DEFAULTS[kind] || TUNNEL_KIND_DEFAULTS.awg;
        this.kind = data.kind || kind;
        this.id = data.id || 0;
        this.enable = data.enable || false;
        this.interfaceName = data.interfaceName || d.interfaceName;
        this.listenPort = data.listenPort || d.listenPort;
        this.mtu = data.mtu || 1420;
        this.privateKey = data.privateKey || '';
        this.publicKey = data.publicKey || '';
        this.ipv4Address = data.ipv4Address || d.ipv4Address;
        this.ipv4Pool = data.ipv4Pool || d.ipv4Pool;
        this.ipv6Enabled = data.ipv6Enabled || false;
        this.ipv6Address = data.ipv6Address || '';
        this.ipv6Pool = data.ipv6Pool || '';
        this.ipv6Gateway = data.ipv6Gateway || '';

        if (d.obfuscation) {
            this.jc = data.jc !== undefined ? data.jc : 4;
            this.jmin = data.jmin !== undefined ? data.jmin : 50;
            this.jmax = data.jmax !== undefined ? data.jmax : 1000;
            this.s1 = data.s1 !== undefined ? data.s1 : 0;
            this.s2 = data.s2 !== undefined ? data.s2 : 0;
            this.s3 = data.s3 !== undefined ? data.s3 : 0;
            this.s4 = data.s4 !== undefined ? data.s4 : 0;
            // H1-H4 are strings: a single value ("1") for 1.x or a "low-high" range for 2.0.
            this.h1 = data.h1 !== undefined ? String(data.h1) : '1';
            this.h2 = data.h2 !== undefined ? String(data.h2) : '2';
            this.h3 = data.h3 !== undefined ? String(data.h3) : '3';
            this.h4 = data.h4 !== undefined ? String(data.h4) : '4';
            this.i1 = data.i1 !== undefined ? data.i1 : '';
            this.i2 = data.i2 !== undefined ? data.i2 : '';
            this.i3 = data.i3 !== undefined ? data.i3 : '';
            this.i4 = data.i4 !== undefined ? data.i4 : '';
            this.i5 = data.i5 !== undefined ? data.i5 : '';
            // AmneziaWG 3.0. Empty means "not written to the config", so the
            // kernel keeps its default and a 2.0 server stays exactly as it was.
            this.headerProtectionKey = data.headerProtectionKey || '';
            this.contentPaddingAddition = data.contentPaddingAddition || '';
            this.rekeyAfterTime = data.rekeyAfterTime || '';
            this.rekeyTimeout = data.rekeyTimeout || '';
            this.rejectAfterTime = data.rejectAfterTime || '';
            this.keepaliveTimeout = data.keepaliveTimeout || '';
            this.maxHandshakeAttempts = data.maxHandshakeAttempts || '';
            this.randomTrailers = data.randomTrailers || false;
            this.disableCookies = data.disableCookies || false;
        }

        this.dnsIpv4 = data.dnsIpv4 || '1.1.1.1';
        this.dnsIpv6 = data.dnsIpv6 || '2606:4700:4700::1111';
        this.externalInterface = data.externalInterface || '';
        this.ipv6ExternalInterface = data.ipv6ExternalInterface || undefined;
        this.postUp = data.postUp || '';
        this.postDown = data.postDown || '';
        this.endpoint = data.endpoint || '';
        this.trafficReset = data.trafficReset || 'never';
        this.routeViaXray = data.routeViaXray || false;
        this.xrayInboundTag = data.xrayInboundTag || d.xrayInboundTag;
        this.xrayTproxyPort = data.xrayTproxyPort || d.xrayTproxyPort;
    }
}

// Clients are identical for both flavours.
class TunnelClient {
    constructor(data = {}) {
        this.id = data.id || 0;
        this.serverId = data.serverId || 0;
        this.name = data.name || '';
        this.email = data.email || '';
        this.enable = data.enable !== undefined ? data.enable : true;
        this.comment = data.comment || '';
        this.privateKey = data.privateKey || '';
        this.publicKey = data.publicKey || '';
        this.presharedKey = data.presharedKey || '';
        this.ipv4Address = data.ipv4Address || '';
        this.ipv6Address = data.ipv6Address || '';
        this.allowedIPs = data.allowedIPs || '';
        this.clientAllowedIPs = data.clientAllowedIPs || '0.0.0.0/0, ::/0';
        this.forwardedPorts = data.forwardedPorts || '';
        this.persistentKeepalive = data.persistentKeepalive !== undefined ? data.persistentKeepalive : 25;
        this.upload = data.upload || 0;
        this.download = data.download || 0;
        this.totalGB = data.totalGB || 0;
        this.allTime = data.allTime || 0;
        this.expiryTime = data.expiryTime || 0;
        this.reset = data.reset || 0;
        this.limitIp = data.limitIp || 0;
        this.tgId = data.tgId || 0;
        this.lastOnline = data.lastOnline || 0;
        this.lastIp = data.lastIp || '';
        this.createdAt = data.createdAt || 0;
        this.updatedAt = data.updatedAt || 0;
    }
}

// Flavour-named wrappers so existing call sites keep reading naturally.
class AwgServer extends TunnelServer {
    constructor(data = {}) { super('awg', data); }
}

class WgServer extends TunnelServer {
    constructor(data = {}) { super('wg', data); }
}

class AwgClient extends TunnelClient {}
class WgClient extends TunnelClient {}
